### DISCUSSION TITLE
uv_write_cb error handling

### DISCUSSION THREAD

**Discussion by chenziliang** [libuv/libuv#4267] [Q&A]
Hi,

In my current code, I tried to do some error handling in `uv_write_cb`. Specifically, what i did is check the `status` in `uv_write_cb`, if the status indicates a permanent error , call `uv_close(handle, closeCallback)` to clean up handle.

The issue I run into is when an EPIPE status is reported in `uv_write_cb`, calling uv_close basically trigger assert failure : the handle is already closing, which seems libuv will close the handle when something really bad happened. This makes me wonder, when will libuv close the handle automatically and if libuv close the handle automatically in some scenarios, how does libuv notify application it is closing the handle.

Thanks a lot !

-----

My code snippet 

```
uv_write(..., writeCallback);

void writeCallback(uv_write_t * req, int status)
{
    if (status != 0)
    {
        if (isPermanentError(status)
            uv_close(req->handle, closeCallback); <--- This line trigger assert failure in debug mode: assert(!uv__is_closing(handle)); 
    }
}

```

**Comment by chenziliang** [comment]
I closed twice in my app

**Comment by chenziliang** [comment]
Checing the code it seems libuv doesn't close handle automatically

**Reply by bnoordhuis** [reply]
That's right.
